### PR TITLE
Update documentation to highlight default use of serializable isolation

### PR DESCRIPTION
**Pull request by spencerkimball**
Deprecate snapshot isolation, and rewrite some explanations as necessary
to make it clear why serializable is a necessary default.

### CODE REVIEW

**Comment by bdarnell** [comment]
:lgtm:

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[v2.0/transactions.md, line 226 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2537#-L5Z3tX2a0IditssHUXl:-L5Z3tX2a0IditssHUXm:bh9tfsv) ([raw file](https://github.com/cockroachdb/docs/blob/bf268037c88b3c358ceaae56c43c4f890a4f1ccd/v2.0/transactions.md#L226)):*
> ```Markdown
> #### Snapshot Isolation
> 
> With `SNAPSHOT` isolation (**deprecated**), a transaction behaves as if it were reading the state of the database consistently at a fixed point in time. Unlike the `SERIALIZABLE` level, `SNAPSHOT` isolation permits the write skew anomaly.
> ```

Do we want to specifically call out that write skew can affect e.g. foreign key checks? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2537#-:-L5Z5GncypiuzyciL3Bg:bnfp4nl)*

**Comment by jseldess** [comment]
Thanks for these updates, @spencerkimball! Getting on a plane in a few hours. I'll find time later or tomorrow to look closely. 

But I have a basic question: What exactly do we mean by **deprecated**? Are we indicating that it is no longer supported (you'll get an error if you try to use it) or that it's no longer necessary?

**Comment by spencerkimball** [comment]
Good question. I will explain further in the text.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[v2.0/transactions.md, line 226 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/2537#-L5Z3tX2a0IditssHUXl:-L5_A-8QU5FvPsDNVsqu:b-sui4v4) ([raw file](https://github.com/cockroachdb/docs/blob/bf268037c88b3c358ceaae56c43c4f890a4f1ccd/v2.0/transactions.md#L226)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Do we want to specifically call out that write skew can affect e.g. foreign key checks? 
</blockquote></details>

I'd rather specifically mention this because there are other issues as well. Instead I'll make the deprecation more forceful in conjunction with a response to Jesse's comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2537#-:-L5_929ZMjwlGxZ07YUO:b-p9z10j)*

File: v2.0/begin-transaction.md [hunk 1]
```diff
@@ -27,16 +27,15 @@ In CockroachDB, the following are aliases for the `BEGIN` statement:
...
+- `READ UNCOMMITTED`, `READ COMMITTED`, and `REPEATABLE READ` are aliases for `SERIALIZABLE`
 
 For more information on isolation level aliases, see [Comparison to ANSI SQL Isolation Levels](transactions.html#comparison-to-ansi-sql-isolation-levels).
 
 ## Parameters
 
 | Parameter | Description |
 |-----------|-------------|
-| `ISOLATION LEVEL` | If you don't want the transaction to run as `SERIALIZABLE` (CockroachDB's default, which provides the highest level of isolation), you can specify `SNAPSHOT`, which can provide better performance in high-contention scenarios.<br/><br/>For more information, see [Transactions: Isolation Levels](transactions.html#isolation-levels).<br/><br/>**Default**: `SERIALIZABLE` |
+| `ISOLATION LEVEL` | If you don't want the transaction to run as `SERIALIZABLE` (CockroachDB's default, which provides the highest level of isolation), you can specify `SNAPSHOT` (**deprecated**), which can provide fewer restarts in high-contention scenarios, but can cause inconsistencies.<br/><br/>For more information, see [Transactions: Isolation Levels](transactions.html#isolation-levels).<br/><br/>**Default**: `SERIALIZABLE` |
```
Comment (jseldess) [inline v2.0/begin-transaction.md]: The description of `SNAPSHOT` in the transactions doc is much more direct.  Should we follow suit here, for example:

> By default, transactions in CockroachDB implement the strongest ISO isolation level: `SERIALIZABLE`. At this isolation level, transactions will never result in anomalies. The `SNAPSHOT` isolation level is still supported as well for backwards compatibility, but you should avoid using it. It provides little benefit in terms of performance and can result in inconsistent state under certain complex workloads.
>
> For more information, see [Transactions: Isolation Levels](transactions.html#isolation-levels).
>
>**Default**: `SERIALIZABLE`

File: v2.0/frequently-asked-questions.md [hunk 2]
```diff
@@ -100,7 +100,7 @@ Yes. Every [transaction](transactions.html) in CockroachDB guarantees [ACID sema
...
-- **Isolation:** By default, transactions in CockroachDB use serializable snapshot isolation (SSI). This means that even concurrent read-write transactions will never result in anomalies. We also provide snapshot isolation (SI), which is more performant with high-contention workloads, although it exhibits anomalies not present in SSI (write skew). For a detailed discussion of isolation in CockroachDB transactions, see [Serializable, Lockless, Distributed: Isolation in CockroachDB](https://www.cockroachlabs.com/blog/serializable-lockless-distributed-isolation-cockroachdb/).
+- **Isolation:** Transactions in CockroachDB implement the strongest ISO isolation level: serializable (SERIALIZABLE). This means that transactions will never result in anomalies. For a detailed discussion of isolation in CockroachDB transactions, see [Serializable, Lockless, Distributed: Isolation in CockroachDB](https://www.cockroachlabs.com/blog/serializable-lockless-distributed-isolation-cockroachdb/).
```
Comment (jseldess) [inline v2.0/frequently-asked-questions.md]: nit: Let's put back ticks around the `SERIALIZABLE` in parentheses.

Comment (jseldess) [reply v2.0/frequently-asked-questions.md]: Also, should we still link to the Serializable blog post or rather link to the Isolation section of the transactions doc?

File: v2.0/set-transaction.md [hunk 3]
```diff
@@ -22,7 +22,7 @@ No [privileges](privileges.html) are required to set the transaction isolation l
 
 | Parameter | Description |
 |-----------|-------------|
-| `ISOLATION LEVEL` | If you don't want the transaction to run as `SERIALIZABLE` (CockroachDB's default, which provides the highest level of isolation), you can set the isolation level to `SNAPSHOT`, which can provide better performance in high-contention scenarios.<br/><br/>For more information, see [Transactions: Isolation Levels](transactions.html#isolation-levels).<br/><br/>**Default**: `SERIALIZABLE` |
+| `ISOLATION LEVEL` | If you don't want the transaction to run as `SERIALIZABLE` (CockroachDB's default, which provides the highest level of isolation), you can specify `SNAPSHOT` (**deprecated**), which can provide fewer restarts in high-contention scenarios, but can cause inconsistencies.<br/><br/>For more information, see [Transactions: Isolation Levels](transactions.html#isolation-levels).<br/><br/>**Default**: `SERIALIZABLE` |
```
Comment (jseldess) [inline v2.0/set-transaction.md]: See comment for `begin-transaction.md`.

File: v2.0/transactions.md [hunk 4]
```diff
@@ -199,7 +199,7 @@ The client can also display the current priority of the transaction with [`SHOW
 
 ### Isolation Levels
 
-CockroachDB supports two transaction isolation levels: `SERIALIZABLE` and `SNAPSHOT`. By default, transactions use the `SERIALIZABLE` isolation level, but the client can explicitly set a transaction's isolation when starting the transaction:
+CockroachDB efficiently supports the strongest ANSI transaction isolation level: `SERIALIZABLE`. All other ANSI transaction isolaton levels (e.g. `READ UNCOMMITTED`, `READ COMMITTED`, and `REPEATABLE READ`) are automatically upgraded to `SERIALIZABLE`. Weaker isolation levels have historically been used to maximize transaction throughput. However, [recent research](http://www.bailis.org/papers/acidrain-sigmod2017.pdf) has demonstrated that the use of weak isolation levels results in substantial vulnerability to concurrency-based attacks. CockroachDB continues to support an additional non-ANSI isolation level, `SNAPSHOT`, although it is deprecated. Clients can explicitly set a transaction's isolation when starting the transaction:
```
Comment (jseldess) [inline v2.0/transactions.md]: nit: `e.g.,`

File: v2.0/transactions.md [hunk 5]
```diff
@@ -219,27 +219,24 @@ The client can also display the current isolation level of the transaction with
 
 With `SERIALIZABLE` isolation, a transaction behaves as though it has the entire database all to itself for the duration of its execution. This means that no concurrent writers can affect the transaction unless they commit before it starts, and no concurrent readers can be affected by the transaction until it has successfully committed. This is the strongest level of isolation provided by CockroachDB and it's the default.
 
-Unlike `SNAPSHOT`, `SERIALIZABLE` isolation permits no anomalies. However, due to CockroachDB's transaction model, `SERIALIZABLE` isolation may require more transaction restarts, especially in the presence of high contention between concurrent transactions. Consider using `SNAPSHOT` isolation for high contention workloads.
+Unlike `SNAPSHOT`, `SERIALIZABLE` isolation permits no anomalies. `SERIALIZABLE` isolation may require transaction restarts to prevent [write skew](https://en.wikipedia.org/wiki/Snapshot_isolation) anomalies.
```
Comment (jseldess) [inline v2.0/transactions.md]: Elsewhere, we say:

> Unlike the `SERIALIZABLE` level, `SNAPSHOT` isolation permits the write skew anomaly. 

Here we suggest that write skew can happen with `SERIALIZABLE`, which feels contradictory?

File: v2.0/begin-transaction.md (hunk 1 above)
Comment (spencerkimball) [reply v2.0/begin-transaction.md]: Done.

File: v2.0/frequently-asked-questions.md (hunk 2 above)
Comment (spencerkimball) [reply v2.0/frequently-asked-questions.md]: Done.

File: v2.0/set-transaction.md (hunk 3 above)
Comment (spencerkimball) [reply v2.0/set-transaction.md]: Done.

File: v2.0/transactions.md (hunk 4 above)
Comment (spencerkimball) [reply v2.0/transactions.md]: Done.

File: v2.0/transactions.md (hunk 5 above)
Comment (spencerkimball) [reply v2.0/transactions.md]: Re-worded to make this clear. It prevents `write skew` by noticing it's possible and requiring restarts.

**Comment by jseldess** [comment]
Thanks, @spencerkimball! Just noticed that we should update this part of the architecture docs: https://www.cockroachlabs.com/docs/stable/architecture/transaction-layer.html#isolation-levels. 

Would you mind doing that before merging?

Fixes #2526.

**Comment by spencerkimball** [comment]
@jseldess updated `transaction-layer.html`

**Comment by jseldess** [comment]
Thanks, @spencerkimball. Merging.

**Pull request merged**